File: mobile/android/app/src/main/java/com/volttracker/obdpoc/ObdSessionLog.java
package com.volttracker.obdpoc;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Writes the per-session {@code .jsonl} field log to disk: one JSON line per command, event,
 * telemetry sample or status change. Extracted from {@link ObdService} so the file IO is isolated
 * from the OBD/session orchestration. This class owns only the file — the database session
 * lifecycle stays in {@code ObdService}.
 *
 * <p>The writer holds onto the underlying {@link FileOutputStream} so {@link #writeDurable} can
 * call {@code getFD().sync()} after flush — that fsync is what lets the on-disk record of an
 * unexpected process death (error rows in particular) survive a crash or sudden power loss.
 */
final class ObdSessionLog {

    private final File logsDir;
    private FileOutputStream fos;
    private BufferedWriter writer;
    private File file;

    ObdSessionLog(File logsDir) {
        this.logsDir = logsDir;
    }

    /** Opens a fresh {@code session-<ts>-<mode>.jsonl}. Leaves the log closed on failure. */
    synchronized void open(String mode) {
        close();
        if (!logsDir.exists() && !logsDir.mkdirs()) {
            return;
        }
        File pending =
                new File(logsDir, "session-" + System.currentTimeMillis() + "-" + mode + ".jsonl");
        try {
            // Append mode: a re-open within the same millisecond + mode shouldn't truncate prior
            // content, even though we expect a fresh file in practice.
            fos = new FileOutputStream(pending, true);
            writer = new BufferedWriter(new OutputStreamWriter(fos, StandardCharsets.UTF_8));
            file = pending;
            writeLatestPointer(pending);
        } catch (IOException ex) {
            closeQuietly();
        }
    }

    synchronized boolean isOpen() {
        return writer != null;
    }

    /** The current log file name, or {@code null} when the log is closed. */
    synchronized String fileName() {
        return file == null ? null : file.getName();
    }

    /** Appends one {@code {ts,type,payload,file}} line. A no-op when the log is closed. */
    synchronized void write(String type, JSONObject payload) {
        writeInternal(type, payload, false);
    }

    /**
     * Same as {@link #write} but additionally fsyncs the underlying file descriptor so the line
     * survives a process death or sudden power loss. Use for {@code "error"} rows and other
     * "must-not-lose" diagnostics — regular telemetry stays on the buffered path so the poll loop
     * isn't blocked on disk on every sample.
     */
    synchronized void writeDurable(String type, JSONObject payload) {
        writeInternal(type, payload, true);
    }

    private void writeInternal(String type, JSONObject payload, boolean durable) {
        if (writer == null) {
            return;
        }
        JSONObject line = new JSONObject();
        try {
            line.put("ts", System.currentTimeMillis());
            line.put("type", type);
            line.put("payload", payload);
            if (file != null) {
                line.put("file", file.getName());
            }
            writer.write(line.toString());
            writer.newLine();
            writer.flush();
            if (durable && fos != null) {
                // flush() drains the BufferedWriter into the OutputStream and on to the
                // FileOutputStream's underlying FD — but the FD's bytes are still in the OS page
                // cache. sync() blocks until they hit stable storage.
                try {
                    fos.getFD().sync();
                } catch (IOException ignored) {
                    // Already past flush; nothing more we can do here.
                }
            }
        } catch (IOException | JSONException ignored) {
        }
    }

    synchronized void close() {
        closeQuietly();
    }

    private void closeQuietly() {
        if (writer != null) {
            try {
                writer.flush();
                writer.close();
            } catch (IOException ignored) {
            }
        } else if (fos != null) {
            try {
                fos.close();
            } catch (IOException ignored) {
            }
        }
        writer = null;
        fos = null;
        file = null;
    }

    private void writeLatestPointer(File logFile) {
        File pointer = new File(logsDir, "latest.txt");
        try (BufferedWriter pointerWriter = new BufferedWriter(new FileWriter(pointer, false))) {
            pointerWriter.write(logFile.getName());
            pointerWriter.newLine();
        } catch (IOException ignored) {
        }
    }
}

File: mobile/android/app/src/main/kotlin/com/volttracker/obdpoc/ObdSessionLog.kt
package com.volttracker.obdpoc

import org.json.JSONException
import org.json.JSONObject
import java.io.BufferedWriter
import java.io.File
import java.io.FileOutputStream
import java.io.FileWriter
import java.io.IOException
import java.io.OutputStreamWriter
import java.nio.charset.StandardCharsets

/**
 * Writes the per-session `.jsonl` field log to disk.
 */
class ObdSessionLog(
    private val logsDir: File,
) {
    private var fos: FileOutputStream? = null
    private var writer: BufferedWriter? = null
    private var file: File? = null
    private var lastFailure: String? = null
    private var bufferedTelemetryLines = 0
    private var lastFlushAtMs = 0L

    /** Opens a fresh `session-<ts>-<mode>.jsonl`. Leaves the log closed on failure. */
    @Synchronized
    fun open(mode: String) {
        close()
        if (!logsDir.exists() && !logsDir.mkdirs()) {
            noteFailure("open_mkdir_failed", null)
            return
        }
        if (!logsDir.isDirectory) {
            noteFailure("open_mkdir_failed", null)
            return
        }
        val pending = File(logsDir, "session-${System.currentTimeMillis()}-$mode.jsonl")
        try {
            fos = FileOutputStream(pending, true)
            writer = BufferedWriter(OutputStreamWriter(fos, StandardCharsets.UTF_8))
            file = pending
            lastFailure = null
            bufferedTelemetryLines = 0
            lastFlushAtMs = System.currentTimeMillis()
            writeLatestPointer(pending)
        } catch (ex: IOException) {
            noteFailure("open_failed", ex)
            closeQuietly()
        }
    }

    @Synchronized
    fun isOpen(): Boolean = writer != null

    /** Most recent local log IO failure, or null once a later open/write succeeds. */
    @Synchronized
    fun lastWriteFailure(): String? = lastFailure

    /** The current log file name, or `null` when the log is closed. */
    @Synchronized
    fun fileName(): String? = file?.name

    /** Appends one `{ts,type,payload,file}` line. A no-op when the log is closed. */
    @Synchronized
    fun write(
        type: String,
        payload: JSONObject,
    ) {
        writeInternal(type, payload, false)
    }

    @Synchronized
    fun writeDurable(
        type: String,
        payload: JSONObject,
    ) {
        writeInternal(type, payload, true)
    }

    private fun writeInternal(
        type: String,
        payload: JSONObject,
        durable: Boolean,
    ) {
        val currentWriter = writer ?: return
        val line = JSONObject()
        try {
            line.put("ts", System.currentTimeMillis())
            line.put("type", type)
            line.put("payload", payload)
            file?.let { line.put("file", it.name) }
            currentWriter.write(line.toString())
            currentWriter.newLine()
            val now = System.currentTimeMillis()
            if (type == "telemetry" && !durable) bufferedTelemetryLines += 1
            val shouldFlush =
                durable ||
                    type != "telemetry" ||
                    bufferedTelemetryLines >= TELEMETRY_FLUSH_BATCH_LINES ||
                    now - lastFlushAtMs >= TELEMETRY_FLUSH_MAX_DELAY_MS
            if (shouldFlush) {
                currentWriter.flush()
                bufferedTelemetryLines = 0
                lastFlushAtMs = now
            }
            var syncFailedThisCall = false
            var syncSucceededThisCall = false
            if (durable) {
                try {
                    fos?.fd?.sync()
                    syncSucceededThisCall = true
                } catch (ignored: IOException) {
                    syncFailedThisCall = true
                    noteFailure("sync_failed", ignored)
                }
            }
            // A successful flush clears a prior failure — but a stale "sync_failed" may only be
            // cleared by a later durable write whose own sync just succeeded. An unflushed
            // telemetry write proves nothing about disk health and must not mask either failure.
            // (And never clear when THIS call's own fsync just failed — its noteFailure must stand.)
            val staleFailureIsSyncRelated = lastFailure?.startsWith("sync_failed") == true
            if (lastFailure != null &&
                shouldFlush &&
                !syncFailedThisCall &&
                (!staleFailureIsSyncRelated || syncSucceededThisCall)
            ) {
                lastFailure = null
            }
        } catch (ex: IOException) {
            noteFailure("write_failed", ex)
        } catch (ex: JSONException) {
            // The line is dropped, but record it so lastWriteFailure() surfaces
            // the loss instead of the log silently missing entries.
            noteFailure("encode_failed", ex)
        }
    }

    @Synchronized
    fun close() {
        closeQuietly()
    }

    private fun closeQuietly() {
        val currentWriter = writer
        val currentFos = fos
        if (currentWriter != null) {
            try {
                currentWriter.flush()
                currentWriter.close()
            } catch (ex: IOException) {
                OBDLog.warn("ObdSessionLog", "close_failed: ${ex.javaClass.simpleName}: ${ex.message}")
            }
        } else if (currentFos != null) {
            try {
                currentFos.close()
            } catch (ex: IOException) {
                OBDLog.warn("ObdSessionLog", "close_failed: ${ex.javaClass.simpleName}: ${ex.message}")
            }
        }
        writer = null
        fos = null
        file = null
        bufferedTelemetryLines = 0
        lastFlushAtMs = 0L
    }

    private fun writeLatestPointer(logFile: File) {
        val pointer = File(logsDir, "latest.txt")
        try {
            BufferedWriter(FileWriter(pointer, false)).use { pointerWriter ->
                pointerWriter.write(logFile.name)
                pointerWriter.newLine()
            }
        } catch (ignored: IOException) {
            noteFailure("latest_pointer_failed", ignored)
        }
    }

    private fun noteFailure(
        stage: String,
        ex: Exception?,
    ) {
        val detail = if (ex == null) stage else "$stage: ${ex.javaClass.simpleName}: ${ex.message}"
        lastFailure = detail
        OBDLog.warn("ObdSessionLog", detail)
    }

    companion object {
        const val TELEMETRY_FLUSH_BATCH_LINES: Int = 8
        const val TELEMETRY_FLUSH_MAX_DELAY_MS: Long = 5_000L
    }
}
